### ISSUE TITLE
[Package]: openjfx

### ISSUE THREAD

**Issue by masterjavaofficial** [termux/termux-packages#7464]
### Package description

Rich client application platform for Java

### Home page URL

https://openjfx.io

### Source code URL

https://github.com/openjdk/jfx

### Packaging policy acknowledgement

- [X] I have read and understand the [Packaging Policy](https://github.com/termux/termux-packages/blob/master/CONTRIBUTING.md#packaging-policy).

### Additional information

![Screenshot_2021-09-03-13-51-55-725_com termux](https://user-images.githubusercontent.com/78401450/132030467-861c67de-c816-4980-b9ce-c7e7c1a3692d.jpg)

I tried to build the package using Termux from my own device (armv71 architecture), but it seems that arm is not supported for building it. Rather, it supports cross-compiling, so it can be built on another Linux architecture targeting my device architecture and others.

## More Information
JavaFX/OpenJFX is the latest GUI application development toolkit for OpenJDK. JavaFX was included in Java SE till JDK 8. Afterwards, to speed up the development of JavaFX, it was excluded from Java SE and open sourced as OpenJFX, and now developed separately (but still under the management of Oracle). As Java's older GUI toolkits such as AWT/Swing had many limitations and didn't support many modern UI/UX APIs, here JavaFX provides many rich APIs, including hardware acceleration and so on, so its now most popular.

Now, as OpenJDK 17 is available on Termux as ```openjdk-17``` package, it is right time to add OpenJFX as a package also. OpenJDK 17 can run OpenJFX 17 & 18 versions (OpenJFX N is compatible with OpenJDK N & N-1). If you can, add OpenJFX 18, please add it. Otherwise 17 will also be also OK.

Ubuntu and other popular Linux distributions provides OpenJFX as ```openjfx``` package. I coded and built one JavaFX application in proot-distro ```ubuntu``` (hirsute), using packages ```openjdk-16-jdk``` & ```openjfx``` there, and was able to run perfectly on Termux’s X11 window system (not the ```proot-distro's```) with XFCE4 using ```xhost``` (```xorg-xhosts```).

**Comment by amrmr739992649** [comment]
> ### Package description
> Rich client application platform for Java
> 
> ### Home page URL
> https://openjfx.io
> 
> ### Source code URL
> https://github.com/openjdk/jfx
> 
> ### Building instructions
> https://wiki.openjdk.java.net/display/OpenJFX/Building+OpenJFX
> 
> ### Packaging policy acknowledgement
> * [x]  I have read and understand the [Packaging Policy](https://github.com/termux/termux-packages/blob/master/CONTRIBUTING.md#packaging-policy).
> 
> ### Additional information
> ![Screenshot_2021-09-03-13-51-55-725_com termux](https://user-images.githubusercontent.com/78401450/132030467-861c67de-c816-4980-b9ce-c7e7c1a3692d.jpg)
> 
> I tried to build the package using Termux from my own device (armv7l architecture), but it seems that arm is not supported for building it. Rather, it supports cross-compiling, so it can be built on another Linux architecture targeting my device architecture and others.
> 
> ### More information
> JavaFX/OpenJFX is the latest GUI application development toolkit for OpenJDK. JavaFX was included in Java SE till JDK 8. Afterwards, to speed up the development of JavaFX, it was excluded from Java SE and open sourced as OpenJFX, and now developed separately (but still under the management of Oracle). As Java's older GUI toolkits such as AWT/Swing had many limitations and didn't support many modern UI/UX APIs, here JavaFX provides many rich APIs, including hardware acceleration and so on, so its now most popular.
> 
> Now, as OpenJDK 17 is available on Termux as `openjdk-17` package, it is right time to add OpenJFX as a package also. OpenJDK 17 can run OpenJFX 17 & 18 versions (OpenJFX N is compatible with OpenJDK N & N-1). If you can add OpenJFX 18, please add it. Otherwise 17 will also be also OK.
> 
> Ubuntu and other popular Linux distributions provides OpenJFX as `openjfx` package. I coded and built one JavaFX application in `proot-distro` `ubuntu` (hirsute), using packages `openjdk-16-jdk` & `openjfx` there, and was able to run perfectly on Termux’s X11 window system (not the `proot-distro`'s) with XFCE4 & VNC Viewer app using `xhost` (`xorg-xhosts`).

**Comment by suhan-paradkar** [comment]
I guess this line needs to be patched in the source code tho
https://github.com/openjdk/jfx/blob/f987b181aba1d3d56e745bd02fe350287ab57459/build.gradle#L311

**Comment by amrmr739992649** [comment]
https://github.com/openjdk/jfx/blob/f987b181aba1d3d56e745bd02fe350287ab57459/build.gradle#L311

**Comment by masterjavaofficial** [comment]
> I guess this line needs to be patched in the source code tho
> https://github.com/openjdk/jfx/blob/f987b181aba1d3d56e745bd02fe350287ab57459/build.gradle#L311

Humm. Other many things needs to be patched, including native libraries.

**Comment by masterjavaofficial** [comment]
@xeffyr @Grimler91 I'm building this package on my arm32 phone.

While building a native JNI library, I'm facing this error:

```
CANNOT LINK EXECUTABLE "g++": cannot locate symbol "_ZNKSt6__ndk119__shared_weak_count13__get_deleterERKSt9type_info" referenced by "/data/data/com.termux/files/usr/bin/clang-13"
```

Any way to resolve this?

[However, previously I faced another issue when compiling with g++ saying that the dynamic library```c++_shared``` was not found, so I've symlinked ```/system/lib/libc++.so``` as ```$PREFIX/lib/libc++_shared.so```, and it was fixed.]

**Comment by Grimler91** [comment]
@masterjavaofficial what's the `termux-info` output of your device? 

Compiling simple example program with g++ seem to work fine on my android 11 arm tablet, with all packages up to date

**Comment by masterjavaofficial** [comment]
> @masterjavaofficial what's the `termux-info` output of your device?
```
Application version:
0.117
Packages CPU architecture:
arm
Subscribed repositories:
# sources.list
deb https://grimler.se/termux-packages-24 stable main
# science-repo (sources.list.d/science.list)
deb https://grimler.se/science-packages-24 science stable
# game-repo (sources.list.d/game.list)
deb https://grimler.se/game-packages-24 games stable
# x11-repo (sources.list.d/x11.list)
deb https://grimler.se/x11-packages x11 main
Updatable packages:
All packages up to date
Android version:
10
Kernel build information:
Linux localhost 4.9.193-perf-gc285628 #1 SMP PREEMPT Tue Aug 3 22:31:37 CST 2021 armv7l Android
Device manufacturer:
Xiaomi
Device model:
Redmi 8A
```

> Compiling simple example program with g++ seem to work fine on my android 11 arm tablet, with all packages up to date
I was also able to compile properly, until I've hitted this:
```
ln -sfr /system/lib/libc++.so $PREFIX/lib/libc++_shared.so
```

I'm facing this problem only after linking the shared library which is needed by the openjfx's JNI.

**Comment by Grimler91** [comment]
I guess you have overwritten [termux's libc++ then](https://github.com/termux/termux-packages/blob/master/packages/libc%2B%2B/build.sh). Re-install libc++ and use compiler flags like [rpath](https://github.com/termux/termux-packages/blob/ee97e92bfa340d02767e8cdcb00926004bb292be/scripts/build/termux_step_setup_toolchain.sh#L30) to make your program find libc++_shared.so

**Comment by masterjavaofficial** [comment]
> I guess you have overwritten [termux's libc++ then](https://github.com/termux/termux-packages/blob/master/packages/libc%2B%2B/build.sh). Re-install libc++ and use compiler flags like [rpath](https://github.com/termux/termux-packages/blob/ee97e92bfa340d02767e8cdcb00926004bb292be/scripts/build/termux_step_setup_toolchain.sh#L30) to make your program find libc++_shared.so

@Grimler91 now I even can't use ```apt``` to install ```libc++```, because it also says:

```
CANNOT LINK EXECUTABLE "apt": cannot locate symbol "__emutls_get_address" referenced by "/data/data/com.termux/files/usr/lib/libapt-pkg.so"...
```

I need to download the library manually. Any link? 🙂

**Comment by masterjavaofficial** [comment]
> I guess you have overwritten [termux's libc++ then](https://github.com/termux/termux-packages/blob/master/packages/libc%2B%2B/build.sh). Re-install libc++ and use compiler flags like [rpath](https://github.com/termux/termux-packages/blob/ee97e92bfa340d02767e8cdcb00926004bb292be/scripts/build/termux_step_setup_toolchain.sh#L30) to make your program find libc++_shared.so

@Grimler91 can you tell me what I'm doing wrong here? I can see that the library exists. 🙃

```
~/.../javafx.graphics/build $ g++ native/jsl-decora/linux/SSEUtils.obj \
    native/jsl-decora/linux/SSEBlend_MULTIPLYPeer.obj \
    native/jsl-decora/linux/SSEBlend_SRC_OUTPeer.obj \
    native/jsl-decora/linux/SSEBrightpassPeer.obj \
    native/jsl-decora/linux/SSEBlend_HARD_LIGHTPeer.obj \
    native/jsl-decora/linux/SSEBlend_EXCLUSIONPeer.obj \
    native/jsl-decora/linux/SSEDisplacementMapPeer.obj \
    native/jsl-decora/linux/SSEBlend_COLOR_BURNPeer.obj \
    native/jsl-decora/linux/SSEPhongLighting_SPOTPeer.obj \
    native/jsl-decora/linux/SSEBlend_REDPeer.obj \
    native/jsl-decora/linux/SSELinearConvolveShadowPeer.obj \
    native/jsl-decora/linux/SSEBlend_GREENPeer.obj \
    native/jsl-decora/linux/SSEBlend_DARKENPeer.obj \
    native/jsl-decora/linux/SSEBlend_SRC_OVERPeer.obj \
    native/jsl-decora/linux/SSEPhongLighting_DISTANTPeer.obj \
    native/jsl-decora/linux/SSEBlend_LIGHTENPeer.obj \
    native/jsl-decora/linux/SSEBlend_COLOR_DODGEPeer.obj \
    native/jsl-decora/linux/SSEBlend_BLUEPeer.obj \
    native/jsl-decora/linux/SSEBlend_DIFFERENCEPeer.obj \
    native/jsl-decora/linux/SSEBlend_SOFT_LIGHTPeer.obj \
    native/jsl-decora/linux/SSEBlend_SCREENPeer.obj \
    native/jsl-decora/linux/SSEColorAdjustPeer.obj \
    native/jsl-decora/linux/SSELinearConvolvePeer.obj \
    native/jsl-decora/linux/SSEBoxBlurPeer.obj \
    native/jsl-decora/linux/SSEPhongLighting_POINTPeer.obj \
    native/jsl-decora/linux/SSESepiaTonePeer.obj \
    native/jsl-decora/linux/SSEPerspectiveTransformPeer.obj \
    native/jsl-decora/linux/SSEBlend_SRC_INPeer.obj \
    native/jsl-decora/linux/SSEBlend_SRC_ATOPPeer.obj \
    native/jsl-decora/linux/SSEBlend_OVERLAYPeer.obj \
    native/jsl-decora/linux/SSEInvertMaskPeer.obj \
    native/jsl-decora/linux/SSEBlend_ADDPeer.obj \
    native/jsl-decora/linux/SSEBoxShadowPeer.obj \
    -o \
    libs/jsl-decora/linux/libdecora_sse.so \
    -static-libgcc \
    -static-libstdc++ \
    -shared \
    -fno-strict-aliasing \
    -fPIC \
    -fno-omit-frame-pointer \
    -fstack-protector \
    -Wextra \
    -Wall \
    -Wformat-security \
    -Wno-unused \
    -Wno-parentheses \
    -Werror=trampolines \
    -m32 \
    -z \
    relro \
    -Wl \
    # --gc-sections
ld.lld: error: unable to find library -lc++_shared
clang-13: error: linker command failed with exit code 1 (use -v to see invocation)
~/.../javafx.graphics/build $ ls $PREFIX/lib | grep c++_shared
libc++_shared.so # The library actually exists
~/.../javafx.graphics/build $
```

**Comment by Grimler91** [comment]
Add `-Wl,-rpath=$TERMUX_PREFIX/lib` as well and it should work

**Comment by masterjavaofficial** [comment]
@Grimler91 are all these packages available in Termux (maybe with another name)? These are required to build OpenJFX as per said in [here](https://wiki.openjdk.java.net/display/OpenJFX/Building+OpenJFX#BuildingOpenJFX-Ubuntu20.04). Please list the alternatives. 

- libavformat58
- libxxf86vm-dev
- libgl1-mesa-dev
- libx11-dev
- x11proto-core-dev
- libgtk2.0-dev
- libgtk-3-dev

**Comment by xtkoba** [comment]
I guess they are respectively

 * ffmpeg
 * libxxf86vm
 * mesa
 * libx11
 * xorgproto
 * gtk2
 * gtk3

**Comment by masterjavaofficial** [comment]
> I guess they are respectively
> 
> * ffmpeg
> * libxxf86vm
> * mesa
> * libx11
> * xorgproto
> * gtk2
> * gtk3

Are all the required packages/libs covered here?

**Comment by xtkoba** [comment]
`glu` in Termux should correspond to `xlibmesa-glu-dev` in Ubuntu, which is not listed above.

**Comment by masterjavaofficial** [comment]
> `glu` in Termux should correspond to `xlibmesa-glu-dev` in Ubuntu, which is not listed above.

But in another [thread](https://github.com/termux/termux-packages/issues/8005#issuecomment-974598226), where I asked for libgl1-mesa-dev package, you told me that it was named to glu. 😕

**Comment by xtkoba** [comment]
Ah, in the other issue I meant `libglu` or `libglu-dev` was renamed to `glu`. Not `libgl1-mesa-dev`.

**Comment by masterjavaofficial** [comment]
> Ah, in the other issue I meant `libglu` or `libglu-dev` was renamed to `glu`. Not `libgl1-mesa-dev`.

Ohh Okay. Thanks again.

**Comment by masterjavaofficial** [comment]
> Add `-Wl,-rpath=$PREFIX/lib` as well and it should work

@Grimler91 the problem still persists. What to do now? 😔

**Comment by Grimler91** [comment]
> > Add `-Wl,-rpath=$PREFIX/lib` as well and it should work
> 
> @Grimler91 the problem still persists after adding those. What to do now? pensive

Maybe -L$PREFIX/lib as well then, but would help to see exactly what you run and exactly what the error is

**Comment by masterjavaofficial** [comment]
> Maybe -L$PREFIX/lib as well then, but would help to see exactly what you run and exactly what the error is

Still same. What I run is here: https://github.com/termux/termux-packages/issues/7464#issuecomment-966386911

**Comment by Grimler91** [comment]
Maybe issue is `    -static-libstdc++ \`, we don't have a static libc++ library

**Comment by masterjavaofficial** [comment]
> Maybe issue is ` -static-libstdc++ \`, we don't have a static libc++ library

☹️☹️😔🥺

**Comment by masterjavaofficial** [comment]
@xeffyr would you not mind integrate build openjfx with openjdk-17? I'm not able to build it. 🥺

**Comment by masterjavaofficial** [comment]
@xeffyr Would you build and add openjfx to Termux? I'm not able to build it. 😪

**Comment by thunder-coding** [comment]
@masterjavaofficial, I would appreciate if you pull up a draft PR with all your current findings and fixes instead of mentioning random people like that.
